File: src/notices/strategies/skku-standard.ts
import { Fetcher } from '../../shared/fetcher.js';
import { loadHtml, extractText, extractAttr } from '../parser.js';
import { NoticeListItem, NoticeDetail } from '../normalizer.js';
import type { SkkuStandardDepartmentConfig, CrawlStrategy, DetailRef } from '../types.js';
import logger from '../../shared/logger.js';

export class SkkuStandardStrategy implements CrawlStrategy {
  private fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  async crawlList(config: SkkuStandardDepartmentConfig, page: number): Promise<NoticeListItem[]> {
    const offset = page * config.pagination.limit;
    const extra = config.extraParams
      ? Object.entries(config.extraParams).map(([k, v]) => `${k}=${v}`).join('&') + '&'
      : '';
    const url = `${config.baseUrl}?${extra}mode=list&articleLimit=${config.pagination.limit}&${config.pagination.param}=${offset}`;

    logger.info({ url, page }, 'Fetching list page');
    const html = await this.fetcher.fetch(url);
    const $ = loadHtml(html);

    const items: NoticeListItem[] = [];

    $(config.selectors.listItem).each((_i, el) => {
      try {
        const $el = $(el);

        // Category: [행사/세미나] etc.
        const categoryRaw = extractText($el.find(config.selectors.category));
        const category = categoryRaw.replace(/^\[|\]$/g, '');

        // Title + article link
        const $titleLink = $el.find(config.selectors.titleLink);
        const titleRaw = extractText($titleLink);
        const title = titleRaw.trim();
        const href = extractAttr($titleLink, 'href') || '';

        // Extract articleNo from href: ?mode=view&articleNo=135890&...
        // Type H (portal boardId) uses itemId instead of articleNo
        const articleNoMatch = href.match(/articleNo=(\d+)/) || href.match(/itemId=(\d+)/);
        if (!articleNoMatch) {
          logger.warn({ href, title }, 'Could not extract articleNo from href');
          return; // skip this item
        }
        const articleNo = parseInt(articleNoMatch[1], 10);

        // Info list: [No.24662, 작성자, 날짜, 조회수N]
        const $infoItems = $el.find(config.selectors.infoList);
        const infoTexts: string[] = [];
        $infoItems.each((_j, li) => {
          infoTexts.push($(li).text().trim());
        });

        // Parse info fields
        let author: string;
        let date: string;
        let views: number;

        if (config.infoParser === 'labeled') {
          // Label-based meta: "POSTED DATE : 2026-03-20", "WRITER : 화학과", "HIT : 280"
          const infoMap: Record<string, string> = {};
          for (const text of infoTexts) {
            const match = text.match(/^(.+?)\s*:\s*(.+)$/);
            if (match) {
              infoMap[match[1].trim().toUpperCase()] = match[2].trim();
            }
          }
          date = infoMap['POSTED DATE'] || '';
          author = infoMap['WRITER'] || '';
          const hitsText = infoMap['HIT'] || '0';
          const hitsMatch = hitsText.match(/(\d+)/);
          views = hitsMatch ? parseInt(hitsMatch[1], 10) : 0;
        } else {
          // Standard: infoTexts[0]: "No.24662" → extract number part (optional, not used as articleNo)
          author = infoTexts[1] || '';
          date = infoTexts[2] || '';
          const viewsText = infoTexts[3] || '0';
          const viewsMatch = viewsText.match(/(\d+)/);
          views = viewsMatch ? parseInt(viewsMatch[1], 10) : 0;
        }

        items.push({
          articleNo,
          title,
          category,
          author,
          date,
          views,
          detailPath: href,
        });
      } catch (err) {
        logger.warn({ error: (err as Error).message }, 'Failed to parse list item');
      }
    });

    logger.info({ page, count: items.length }, 'Parsed list page');
    return items;
  }

  async crawlDetail(ref: DetailRef, config: SkkuStandardDepartmentConfig): Promise<NoticeDetail | null> {
    // Use detailPath from list page when available (handles itemId/viewBoardId patterns)
    let url: string;
    if (ref.detailPath.startsWith('http')) {
      url = ref.detailPath;
    } else if (ref.detailPath.startsWith('?')) {
      // Inject extraParams (e.g., boardId) if not already present in detailPath
      if (config.extraParams) {
        const params = new URLSearchParams(ref.detailPath.slice(1));
        for (const [k, v] of Object.entries(config.extraParams)) {
          if (!params.has(k)) params.set(k, v);
        }
        url = `${config.baseUrl}?${params.toString()}`;
      } else {
        url = `${config.baseUrl}${ref.detailPath}`;
      }
    } else {
      // Fallback: construct standard SKKU URL
      const extra = config.extraParams
        ? Object.entries(config.extraParams).map(([k, v]) => `${k}=${v}`).join('&') + '&'
        : '';
      url = `${config.baseUrl}?${extra}mode=view&articleNo=${ref.articleNo}&article.offset=0&articleLimit=10`;
    }

    try {
      logger.debug({ url, articleNo: ref.articleNo }, 'Fetching detail page');
      const html = await this.fetcher.fetch(url);
      const $ = loadHtml(html);

      // Content: try configured selector first, then fallback for newer templates
      let $content = $(config.selectors.detailContent);
      if ($content.length === 0) {
        // Department sites (cse, mech, etc.) use div.board-view-content-wrap
        // instead of dl.board-write-box dd used by www.skku.edu
        const fallbacks = ['div.board-view-content-wrap', 'div.fr-view'];
        for (const sel of fallbacks) {
          $content = $(sel);
          if ($content.length > 0) {
            logger.info({ articleNo: ref.articleNo, fallback: sel }, 'Used fallback content selector');
            break;
          }
        }
      }
      const content = $content.html()?.trim() || '';
      const contentText = $content.text().trim();

      // Attachments
      const attachments: { name: string; url: string }[] = [];
      $(config.selectors.attachmentList).each((_i, el) => {
        const $a = $(el);
        const name = extractText($a);

        let fileUrl: string | undefined;
        if (config.attachmentParser === 'onclick') {
          // Parse <button onclick="location.href='...'">
          const onclick = extractAttr($a, 'onclick') || '';
          const onclickMatch = onclick.match(/location\.href='([^']+)'/);
          fileUrl = onclickMatch ? onclickMatch[1].replace(/&amp;/g, '&') : undefined;
        } else {
          fileUrl = extractAttr($a, 'href');
        }

        if (name && fileUrl && fileUrl !== '#') {
          let fullUrl: string;
          if (fileUrl.startsWith('http')) {
            fullUrl = fileUrl;
          } else if (fileUrl.startsWith('?')) {
            fullUrl = `${config.baseUrl}${fileUrl}`;
          } else {
            // Relative path → use origin from baseUrl
            const origin = new URL(config.baseUrl).origin;
            fullUrl = `${origin}${fileUrl.startsWith('/') ? '' : '/'}${fileUrl}`;
          }
          attachments.push({ name, url: fullUrl });
        }
      });

      return { content, contentText, attachments };
    } catch (err) {
      logger.error({ articleNo: ref.articleNo, error: (err as Error).message }, 'Failed to fetch detail page');
      return null;
    }
  }
}

File: py/src/skkuverse_crawler/notices/strategies/skku_standard.py
from __future__ import annotations

import re
from urllib.parse import urljoin, urlparse, urlencode, parse_qs, urlunparse

from ...shared.fetcher import Fetcher
from ...shared.logger import get_logger
from ..models import NoticeDetail, NoticeListItem
from ..parser import load_html, extract_text, extract_attr
from ..types import DetailRef

logger = get_logger("skku_standard")


class SkkuStandardStrategy:
    def __init__(self, fetcher: Fetcher) -> None:
        self.fetcher = fetcher

    async def crawl_list(self, config: dict, page: int) -> list[NoticeListItem]:
        pagination = config["pagination"]
        offset = page * pagination["limit"]
        extra = ""
        if config.get("extraParams"):
            extra = "&".join(f"{k}={v}" for k, v in config["extraParams"].items()) + "&"
        url = f"{config['baseUrl']}?{extra}mode=list&articleLimit={pagination['limit']}&{pagination['param']}={offset}"

        logger.info("fetching_list_page", url=url, page=page)
        html = await self.fetcher.fetch(url)
        soup = load_html(html)

        items: list[NoticeListItem] = []
        selectors = config["selectors"]

        for el in soup.select(selectors["listItem"]):
            try:
                # Category
                cat_el = el.select_one(selectors["category"])
                category_raw = extract_text(cat_el)
                category = category_raw.strip("[]")

                # Title + link
                title_link = el.select_one(selectors["titleLink"])
                title = extract_text(title_link).strip()
                href = extract_attr(title_link, "href") or ""

                # Extract articleNo
                m = re.search(r"articleNo=(\d+)", href) or re.search(r"itemId=(\d+)", href)
                if not m:
                    logger.warning("no_article_no", href=href, title=title)
                    continue
                article_no = int(m.group(1))

                # Info list
                info_items = el.select(selectors["infoList"])
                info_texts = [li.get_text(strip=True) for li in info_items]

                if config.get("infoParser") == "labeled":
                    info_map: dict[str, str] = {}
                    for text in info_texts:
                        lm = re.match(r"^(.+?)\s*:\s*(.+)$", text)
                        if lm:
                            info_map[lm.group(1).strip().upper()] = lm.group(2).strip()
                    date = info_map.get("POSTED DATE", "")
                    author = info_map.get("WRITER", "")
                    hits_text = info_map.get("HIT", "0")
                    hits_m = re.search(r"(\d+)", hits_text)
                    views = int(hits_m.group(1)) if hits_m else 0
                else:
                    author = info_texts[1] if len(info_texts) > 1 else ""
                    date = info_texts[2] if len(info_texts) > 2 else ""
                    views_text = info_texts[3] if len(info_texts) > 3 else "0"
                    views_m = re.search(r"(\d+)", views_text)
                    views = int(views_m.group(1)) if views_m else 0

                items.append(NoticeListItem(
                    articleNo=article_no,
                    title=title,
                    category=category,
                    author=author,
                    date=date,
                    views=views,
                    detailPath=href,
                ))
            except Exception as exc:
                logger.warning("parse_list_item_failed", error=str(exc))

        logger.info("parsed_list_page", page=page, count=len(items))
        return items

    async def crawl_detail(self, ref: DetailRef, config: dict) -> NoticeDetail | None:
        # Build detail URL
        if ref["detailPath"].startswith("http"):
            url = ref["detailPath"]
        elif ref["detailPath"].startswith("?"):
            if config.get("extraParams"):
                params = dict(parse_qs(ref["detailPath"][1:], keep_blank_values=True))
                flat_params = {k: v[0] if isinstance(v, list) else v for k, v in params.items()}
                for k, v in config["extraParams"].items():
                    if k not in flat_params:
                        flat_params[k] = v
                url = f"{config['baseUrl']}?{urlencode(flat_params)}"
            else:
                url = f"{config['baseUrl']}{ref['detailPath']}"
        else:
            extra = ""
            if config.get("extraParams"):
                extra = "&".join(f"{k}={v}" for k, v in config["extraParams"].items()) + "&"
            url = f"{config['baseUrl']}?{extra}mode=view&articleNo={ref['articleNo']}&article.offset=0&articleLimit=10"

        try:
            html = await self.fetcher.fetch(url)
            soup = load_html(html)
            selectors = config["selectors"]

            # Content with fallbacks
            content_el = soup.select_one(selectors["detailContent"])
            if not content_el:
                for fallback in ["div.board-view-content-wrap", "div.fr-view"]:
                    content_el = soup.select_one(fallback)
                    if content_el:
                        break

            content = content_el.decode_contents().strip() if content_el else ""
            content_text = content_el.get_text(strip=True) if content_el else ""

            # Attachments
            attachments: list[dict[str, str]] = []
            for a in soup.select(selectors["attachmentList"]):
                name = extract_text(a)

                if config.get("attachmentParser") == "onclick":
                    onclick = extract_attr(a, "onclick") or ""
                    om = re.search(r"location\.href='([^']+)'", onclick)
                    file_url = om.group(1).replace("&amp;", "&") if om else None
                else:
                    file_url = extract_attr(a, "href")

                if name and file_url and file_url != "#":
                    if file_url.startswith("http"):
                        full_url = file_url
                    elif file_url.startswith("?"):
                        full_url = f"{config['baseUrl']}{file_url}"
                    else:
                        parsed = urlparse(config["baseUrl"])
                        origin = f"{parsed.scheme}://{parsed.netloc}"
                        full_url = f"{origin}{'/' if not file_url.startswith('/') else ''}{file_url}"
                    attachments.append({"name": name, "url": full_url})

            return NoticeDetail(content=content, contentText=content_text, attachments=attachments)
        except Exception as exc:
            logger.error("detail_fetch_failed", articleNo=ref["articleNo"], error=str(exc))
            return None
